ssl/statem/statem_srvr.c

```diff
@@ -4297,21 +4298,107 @@ CON_FUNC_RETURN tls_construct_new_session_ticket(SSL_CONNECTION *s, WPACKET *pkt
  */
-int tls_construct_cert_status_body(SSL_CONNECTION *s, WPACKET *pkt)
+int tls_construct_cert_status_body(SSL_CONNECTION *s, size_t chainidx, WPACKET *pkt)
 {
-    if (!WPACKET_put_bytes_u8(pkt, s->ext.status_type)
-            || !WPACKET_sub_memcpy_u24(pkt, s->ext.ocsp.resp,
-                                       s->ext.ocsp.resp_len)) {
+    unsigned char *respder = NULL;
+    int resplen = 0;
+#ifndef OPENSSL_NO_OCSP
+    int i = 0;
+    X509 *x = NULL;
+    X509 *issuer = NULL;
+    STACK_OF(X509) *chain_certs = NULL;
+    int found = 0;
+    SSL *ssl = SSL_CONNECTION_GET_SSL(s);
+    OCSP_RESPONSE *resp = NULL;
+    OCSP_BASICRESP *bs = NULL;
+    OCSP_CERTID *cert_id = NULL;
+#endif
+
+    if (!WPACKET_put_bytes_u8(pkt, s->ext.status_type)) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+
+#ifndef OPENSSL_NO_OCSP
+    /*
+     * In TLSv1.3 the caller gives the index of the certificate for which the
+     * status message should be created.
+     * Prior to TLSv1.3 the chain index is 0 and the body should only the status
+     * of the server certificate itself.
+     */
+    x = SSL_get_certificate(ssl);
+
+    SSL_get0_chain_certs(ssl, &chain_certs);
+
+    if (chain_certs != NULL) {
+        /*
+         * if the certificate chain was built, get the status message for the
+         * requested certificate specified by chainidx  SSL_get0_chain_certs
+         * contains certificate chain except the server cert
+         *
+         * if chainidx = 0 the server certificate is request
+         * if chainidx > 0 an intermediate certificate is request
+         */
+        if ((int)chainidx < sk_X509_num(chain_certs)+1 && chainidx > 0) {
+            x = sk_X509_value(chain_certs, chainidx - 1);
+        }
+
+        issuer = X509_find_by_subject(chain_certs, X509_get_issuer_name(x));
+
+        /* search the stack for the requested OCSP response */
+        cert_id = OCSP_cert_to_id(NULL, x, issuer);
+
+        /* find the correct OCSP response for the requested certificate */
+        found = -1;
+        for (i = 0; i < sk_OCSP_RESPONSE_num(s->ext.ocsp.resp_ex); i++) {
+            if ((resp = sk_OCSP_RESPONSE_value(s->ext.ocsp.resp_ex, i)) == NULL)
+                continue;
+
+            if ((bs = OCSP_response_get1_basic(resp)) == NULL)
+                continue;
+
+            found = OCSP_resp_find(bs, cert_id, -1);
+
+            if (bs != NULL)
+                OCSP_BASICRESP_free(bs);
+
+            if (found > -1)
+                break;
+        }
+        if (found < 0)
+            resp = NULL;
+
+    } else if (chainidx == 0) {
+        /*
+         * if certificate chain was not built and only the response for the server
+         * certificate (chainidx=0) was requested, check if the stack contains an
+         * OCSP response and get the first one
+         * TODO: check if the first response on the stack is indeed the one for the
```

Comment:
Would it be hard to implement this TODO?

Reply:
From my perspective it is easier to search a stack (with a very limited number of entries) for the correct OCSP response then assure that the stack has always the correct order and includes NULL for certificates where no response is available. 
The only problem is if no chain is available, but in that case an ordered stack would also not help, because there is only one certificate and one response and you have to trust that its the matching.
If the OCSP responses is received from the OCSP responder this should be assured. In the case the response is provided e.g., via command line option this would be a configuration problem.

Reply:
I agree with  @vdukhovni it would be best to get the correct correspondance from the server.
Yet you could keep the less safe and more complicated but more flexible search behavior (only) for the case that the responses are taken from a list of files.

Reply:
You can move the search behavior to the s_server app
and assume here that the responses are in the right order with missing responses being represented by NULL. 
So you can require that already the app makes sure that the list of OCSP responses (or NULL if not available) is as long as the server cert chain, with the existing responses being at the right place in the list.

Reply:
I would keep it as it is. That's the more safe solution as it doesn't depend that the responses are provided in the correct order especially in the file case.
And it also covers the case if we only have the response for the server certificate.

Reply:
From my perspective this is more robust, as it does not depend on the order of OCSP responses in the stack. 
The problem is also that I don't know at this point how the OCSP responses are provided by files or requested from server.